### ISSUE TITLE
add UI for borg key backup

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbase/vorta#1918]
#### The problem
Now and then I get contacted by borg users who have lost the borg key, asking for help with regaining access to their archived data.

Of course there is no way to do that if the key is really lost - that would be a severe security issue if that was possible.

So the only option is to try to recover the deleted key and that is usually rather complicated and often not successful:

- usage of "undelete" tools for local filesystems, hoping the storage device block(s) with the key have not yet been overwritten or discarded.

- sometimes users directly back up onto FUSE-mounted cloud storage. if they don't have file versioning activated there, guess it is impossible to recover a file from a bucket. otoh, having versioning active might be a bad idea concerning space usage and related costs, because borg compact moves around a lot of data.

So, users should be motivated to have a borg key backup.

`borg init` tells them every time to do that.

Does vorta also tell them? Vorta should also actively help them doing that.

#### Requested Solution

Add a UI wrapping `borg key export` (and maybe also `borg key import` later).

Guess the best way would be to use the option creating that html page with the QR code and the key hexdump and then invoke the default browser showing that html file.

**Comment by goebbe** [comment]
In the current version of Vorta it is easy to export the settings on a per-profile base. Doing this, the users are asked if the repo passphrase should be included to the settings ( .json) file. 
Perhaps the corresponding **keys** (i.e. repokeys?) could/ should optionally be added to these setting .json file?

Currently Vorta does not try to "educate" the user about best practice. Perhaps it is a good idea to actively promote to "save the profile settings" including the repo "passphrase" and "export keys" and to store these savely (e.g. password manager) and keep a copy at a save place (i.e. out side your  computer or even building)?

**Comment by ThomasWaldmann** [comment]
@goebbe borg emits these hints at `borg init` time (on stderr or stdout), but maybe vorta does not show them on the GUI (or they disappear too quickly?).

**Comment by goebbe** [comment]
@ThomasWaldmann  exactly.  I tried to test if **Vorta** passes the  information/ warning about the keys to the user.

If I understand correctly, users should be encouraged to make a backup of the borg keyfile. 
If **keyfiles** are used, this is relatively straightforward (if you know where to find the keyfiles). 
If **repokeys** are uses (the default in Vorta), things get more complicated. To my knowledge, backing up the repokey cannot be achieved easily using Vorta. The user has to call borg commands directly. But maybe I miss something.  

By the way, these are the relevant lines from the log files, when initialising, using keyfiles: 
```
2024-03-08 13:22:13,068 - vorta.borg.borg_job - INFO - Key in "/home/goebbe/.config/borg/keys/192_168_0_5__backup_vorta_repo_test_key" created.
2024-03-08 13:22:13,069 - vorta.borg.borg_job - INFO - Keep this key safe. Your data will be inaccessible without it.
```

**Comment by ThomasWaldmann** [comment]
The step to create a key backup is the same no matter whether keyfile or repokey is used and the backup should be made in both cases.

**Comment by goebbe** [comment]
@ThomasWaldmann 
`borg key export ` could be used in both cases (i.e. when using a (local) keyfile OR when using repokey) - so Vorta could use this command anyway.

In the case of a **(local) keyfile**: Would it be sufficient to just copy/ backup the keyfile directly from `/home/user/.config/borg/keys/your_repo_key` or would` borg export key` do additional magic?

Related question: When **restoring, using Vorta** from a different computer, with a **(local) keyfile**: Would it be sufficient to just copy the keyfile to `/home/user/.config/borg/keys/`  or is it necessary to use `borg key import`

**Comment by ThomasWaldmann** [comment]
`borg key export` offers misc. ways how to export a key and you MUST use that.

Same for `borg key import`.

No manual key file copying.

**Comment by goebbe** [comment]
Related: https://github.com/borgbase/vorta/pull/599  from @samu-w

**Comment by freshreplicant** [comment]
I'm  just piggybacking here, I've read a bunch of issues related to key management, especially those relating to keyfile-blake2. I'm using the Flatpak verison of Vorta, and I have located the keys, e.g. at:

`/var/home/username/.var/app/com.borgbase.Vorta/config/borg/keys`

What I don't understand is, if I need to recover from another device, is it enough to go through the usual steps of importing config/settings and then copying and pasting the file back into:

`/var/home/username/.var/app/com.borgbase.Vorta/config/borg/keys`

Or do I need to do actually run `borg key export` when backing the key up, then run `borg key import` to import it? And if so, is `borg key import` the full command or does it require anything special to make it interface with Vorta (especially installed in Flatpak)? How would I run the borg command with the Flatpak version, do I need to install borg separately?

When I look at the Borg documentation for borg key export, I see this:

```
borg key export /path/to/repo > encrypted-key-backup
borg key export --paper /path/to/repo > encrypted-key-backup.txt
borg key export --qr-html /path/to/repo > encrypted-key-backup.html
```

But if my repo is a borgbase remote repo, and I used keyfile, then the key is not stored in the repo, but only in my home directory, so how do I use this command? Replacing `/path/to/repo` with `/path/to/key` does not work, as it tells you it's not a valid repository.

I understand that repokey is the preferred or officially 'supported' way to backup, but since Vorta has the option of using keyfile, I'd like to figure it out. Any advice would be appreciated.

**Comment by JacksonChen666** [comment]
@freshreplicant `/path/to/repo` is correct (for the `REPOSITORY` argument). It can be an SSH connection string. It is never `/path/to/key` for the `REPOSITORY` argument.

You'd have to run `borg` inside the flatpak, or if that doesn't work, make your system `borg` look within that directory (`$BORG_KEYS_DIR`).

**Comment by goebbe** [comment]
@freshreplicant I would recommend using the default "repokey" method when using Vorta - together with a secure (long) passphrase. 
With repokey, the keyfile is saved with the repository.
To my understanding, the keyfile alone is pretty useless, without the passphrase. So, even if your server is hacked and somebody has access to your keyfiles - nobody can decrypt your backups without the passphrase (without excessive resources/ time).

If you use repokey you don't normally need a backup of the keyfile, if e.g. your computer is destroyed, and you want to access your online backup from a different computer. But of course you would need the passphrase.

Borg recommends making a backup of your keyfile anyway. Imagine, for some reason only your keyfile got damaged/ destroyed/ deleted. In this case, it would be helpful to have a backup of your keyfile, somewhere else.

Using the **Flatpak** build of Vorta you can use the following, in the terminal, to export the keyfiles:
`flatpak run --command=borg com.borgbase.Vorta key export /path/repo > ~/encrypted-repo-key-backup.txt`

And using **Flatpak** for a **remote** backup, via ssh:
`flatpak run --command=borg com.borgbase.Vorta key export ***@***.***:22/~/path/repo > ~/encrypted-repo-key-backup.txt`

edit: I tried to get the formatting right. In the preview everything looks as expected, but nevertheless the markdown is ignored after updating, by github.
